src/Task.cc

```diff
@@ -1085,54 +1089,62 @@ static bool is_singlestep_resume(ResumeRequest request) {
 
 TrapReasons Task::compute_trap_reasons() {
   ASSERT(this, stop_sig() == SIGTRAP);
+
   TrapReasons reasons;
-  uintptr_t status = debug_status();
-  reasons.singlestep = (status & DS_SINGLESTEP) != 0;
 
-  if (is_singlestep_resume(how_last_execution_resumed)) {
-    if (is_at_syscall_instruction(this, address_of_last_execution_resume) &&
-        ip() ==
-            address_of_last_execution_resume +
-                syscall_instruction_length(arch())) {
-      // During replay we execute syscall instructions in certain cases, e.g.
-      // mprotect with syscallbuf. The kernel does not set DS_SINGLESTEP when we
-      // step over those instructions so we need to detect that here.
-      reasons.singlestep = true;
-    } else {
-      TrappedInstruction ti =
-        trapped_instruction_at(this, address_of_last_execution_resume);
-      if (ti == TrappedInstruction::CPUID &&
-          ip() == address_of_last_execution_resume +
-                       trapped_instruction_len(TrappedInstruction::CPUID)) {
-        // Likewise we emulate CPUID instructions and must forcibly detect that
-        // here.
-        reasons.singlestep = true;
-        // This also takes care of the did_set_breakpoint_after_cpuid workaround case
-      } else if (ti == TrappedInstruction::INT3 &&
-          ip() == address_of_last_execution_resume +
-                       trapped_instruction_len(TrappedInstruction::INT3)) {
-        // INT3 instructions should also be turned into a singlestep here.
+  const siginfo_t& si = get_siginfo();
+  if (arch() == x86 || arch() == x86_64) {
+    uintptr_t status = x86_debug_status();
+    reasons.singlestep = (status & DS_SINGLESTEP) != 0;
+    if (!reasons.singlestep && is_singlestep_resume(how_last_execution_resumed)) {
```

Comment:
Why did you add the `!reasons.singlestep` condition?

Reply:
The only thing that could happen in any of these branches is setting singlestep to `true`, so it isn't necessary to try any of these things if the kernel already told us it was a single step. I originally had the condition inverted to avoid going through some x86 code paths on aarch64, but then I decided to do it this way.